net/src/client/tcp/actors/jsonrpc.rs

```diff
@@ -20,40 +24,81 @@ use super::Error;
         log::info!("Connecting client to {}", url);
         let (_handle, socket) = TcpSocket::new(url).map_err(|_| Error::InvalidUrl)?;
         let client = Self {
             _handle,
             socket,
-            url: url.to_owned(),
-            retry_connect: false,
-            subscriber: None,
-            subscription_id: None,
+            active_subscriptions: subscriptions,
+            pending_subscriptions: Default::default(),
+            url: String::from(url),
         };
+        log::info!("TCP socket is now connected to {}", url);
 
-        Ok(client.start())
+        Ok(Actor::start(client))
     }
 
-    /// Renew the connection of the client.
-    pub fn reconnect(&mut self) {
-        log::info!("Reconnecting client to {}", self.url);
-        // The .expect is because the creation of the socket might only fail if the url is invalid,
-        // but since this a reconnection, meaning we were able to correctly parse the url before,
-        // then at this point the url should be the same, hence still valid.
-        let (_handle, socket) = TcpSocket::new(self.url.as_ref()).expect("Unexpected error");
+    /// Replace the TCP connection with a fresh new connection.
+    pub fn reconnect(&mut self, ctx: &mut <Self as Actor>::Context) {
+        log::info!("Reconnecting TCP client to {}", self.url);
+        let (_handle, socket) = TcpSocket::new(self.url.as_str())
+            .map_err(|e| log::error!("Reconnection error: {}", e))
+            .expect("TCP socket reconnection should not panic, as the only possible error is malformed URL");
         self._handle = _handle;
         self.socket = socket;
-        self.retry_connect = false;
-        // TODO: re-subscribe
+
+        // Recover active subscriptions
+        let mut active_subscriptions = self
+            .active_subscriptions
+            .lock()
+            .map(|x| x.clone())
+            .expect("Active subscriptions Mutex should never be poisoned");
+        log::debug!(
+            "Trying to recover {} active subscriptions",
+            active_subscriptions.len()
+        );
+        active_subscriptions.iter().for_each(|(_, subscribe)| {
+            log::debug!("Resubscribing {:?}", subscribe.0);
+            ctx.notify(subscribe.clone());
+        });
+
+        // Process pending subscriptions
+        log::debug!(
+            "Trying to process {} pending subscriptions",
+            self.pending_subscriptions.len()
+        );
+        self.pending_subscriptions
+            .iter()
+            .for_each(|(topic, subscribe)| {
+                log::debug!(
+                    "Processing pending subscription for topic {}: {:?}",
+                    topic,
+                    subscribe.0
+                );
+                ctx.notify(subscribe.clone());
+            });
+
+        // Clear up all subscriptions (will be pushed again if they keep failing)
+        active_subscriptions.clear();
```

Comment:
I think this will not clear `self.active_subscriptions`, it will only clear the clone created by `.map(|x| x.clone())`.